File: samlang-ast/src/main/kotlin/samlang/ast/common/Range.kt
package samlang.ast.common

data class Range(val start: Position, val end: Position) {
    operator fun contains(position: Position): Boolean = position in start..end

    operator fun contains(range: Range): Boolean = contains(position = range.start) && contains(position = range.end)

    infix fun union(other: Range): Range =
        Range(start = minOf(a = start, b = other.start), end = maxOf(a = end, b = other.end))

    override fun toString(): String = "$start-$end"

    companion object {
        @JvmField
        val DUMMY: Range = Range(
            start = Position.DUMMY,
            end = Position.DUMMY
        )
    }
}

File: typescript-rewrite/samlang-core/src/ast/common/range.ts
import Position from './position';

export default class Range {
  static readonly DUMMY: Range = new Range(Position.DUMMY, Position.DUMMY);

  constructor(public readonly start: Position, public readonly end: Position) {}

  readonly containsPosition = (position: Position): boolean =>
    this.start.compareTo(position) <= 0 && this.end.compareTo(position) >= 0;

  readonly containsRange = (range: Range): boolean =>
    this.containsPosition(range.start) && this.containsPosition(range.end);

  readonly union = (range: Range): Range => {
    const start = this.start.compareTo(range.start) < 0 ? this.start : range.start;
    const end = this.end.compareTo(range.end) > 0 ? this.end : range.end;
    return new Range(start, end);
  };

  readonly toString = (): string => `${this.start}-${this.end}`;
}
